web_session_allow_public/tests/test_ir_http.py

```diff
@@ -0,0 +1,25 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 LasLabs Inc.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+import mock
+
+from openerp.tests.common import HttpCase
+
+from openerp.addons.base.ir.ir_http import ir_http
+
+
+MODULE_PATH = 'openerp.addons.base.ir.ir_http'
+
+
+class TestIrHttp(HttpCase):
+
+    @mock.patch('%s.request' % MODULE_PATH)
+    def test_public_session(self, request):
+        """ It should allow HTTP authentication on public user """
+        request.session.uid = False
+        with mock.patch.object(
```

Comment:
> You are the mock guy indeed 

I've said it once and will say it again, I ❤️  mock. I'd probably marry it if there weren't laws against that 😛 

> But, did you know you can actually make RPC calls in tests? 

Ah yes, I spent about two hours trying to make this work yesterday. It turns out that the internal XML proxy is authenticating from a lower level than the external. 

Internal RPC calls directly the [`res.users.check`, which explicitly fails on no passwd](https://github.com/odoo/odoo/blob/457a705bf0927e281d5269014f955d3e2a828e11/odoo/addons/base/res/res_users.py#L496). I tried to override this in [one of my earlier attempts](https://github.com/OCA/web/pull/402/commits/1f8162292bf012c3671b3b3fd27553604a76152c), and [you reported bad results](https://github.com/OCA/web/pull/402#issuecomment-239379531). 

We are now overriding in [the ir.http session authentication](https://github.com/OCA/web/pull/402/commits/643d128cd53c9cc60c71072f7f01a23d2db7beec), which it seems the internal XML-RPC completely bypasses for some reason.

My frustration ala code for further elaboration. The last line illustrates why a mock was required:

```
(Pdb) user = self.env.ref('base.public_user')
(Pdb) openerp.service.security.check('test', user, False)
*** AccessDenied: Access denied
(Pdb) openerp.service.security.check('test', user, False)
*** AccessDenied: Access denied
(Pdb) self.env['res.users'].browse(user).check_credentials(False)
(Pdb) 2016-11-22 02:03:43,309 7962 INFO test werkzeug: 10.211.55.1 - - [22/Nov/2016 02:03:43] "POST /longpolling/poll HTTP/1.1" 200 -
(Pdb) super(IrHttp, self)._auth_method_user()
*** SessionExpiredException: Session expired
self._auth_method_public()
*** ValueError: URI connections not allowed
``